llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp

```diff
@@ -7183,6 +7183,57 @@ bool AMDGPULegalizerInfo::legalizeBVHIntersectRayIntrinsic(
   return true;
 }
 
+bool AMDGPULegalizerInfo::legalizeBVHDualIntrinsic(MachineInstr &MI,
+                                                   MachineIRBuilder &B) const {
+  const LLT S32 = LLT::scalar(32);
+  const LLT V2S32 = LLT::fixed_vector(2, 32);
+
+  Register DstReg = MI.getOperand(0).getReg();
+  Register DstOrigin = MI.getOperand(1).getReg();
+  Register DstDir = MI.getOperand(2).getReg();
+  Register NodePtr = MI.getOperand(4).getReg();
+  Register RayExtent = MI.getOperand(5).getReg();
+  Register InstanceMask = MI.getOperand(6).getReg();
+  Register RayOrigin = MI.getOperand(7).getReg();
+  Register RayDir = MI.getOperand(8).getReg();
+  Register Offsets = MI.getOperand(9).getReg();
+  Register TDescr = MI.getOperand(10).getReg();
+
+  if (!AMDGPU::isGFX12Plus(ST)) {
```

Comment:
This should have a subtarget feature. Generation checks are terrible

Reply:
Could probably just remove this and rely on getting a "cannot select" error later - so long as the compiler does not crash.

Reply:
I added subtarget feature and test.

Reply:
I don't really see the need for a SubtargetFeature. It could just be a nicely named helper function in GCNSubtarget. But I guess this way is OK too.